src/inputs/AddressInput/index.tsx

```diff
@@ -0,0 +1,148 @@
+import React, { ReactElement, useState, ChangeEvent, useEffect } from 'react';
+import InputAdornment from '@material-ui/core/InputAdornment';
+import CircularProgress from '@material-ui/core/CircularProgress';
+
+import { trimSpaces } from '../../utils/strings';
+import {
+  addNetworkPrefix,
+  checksumAddress,
+  getAddressWithoutNetworkPrefix,
+  getNetworkPrefix,
+  isChecksumAddress,
+  isValidAddress,
+  isValidEnsName,
+} from '../../utils/address';
+import useThrottle from '../../utils/useThrottle';
+import TextFieldInput, { TextFieldInputProps } from '../TextFieldInput';
+
+type AddressInputProps = {
+  name: string;
+  address: string;
+  networkPrefix?: string;
+  showNetworkPrefix?: boolean;
+  defaultValue?: string;
+  disabled?: boolean;
+  onChangeAddress: (address: string) => void;
+  getAddressFromDomain?: (name: string) => Promise<string>;
+  showLoadingSpinner?: boolean;
+} & TextFieldInputProps;
+
+function AddressInput({
+  name,
+  address,
+  networkPrefix,
+  showNetworkPrefix,
+  disabled,
+  onChangeAddress,
+  getAddressFromDomain,
+  showLoadingSpinner,
+  InputProps,
+  ...rest
+}: AddressInputProps): ReactElement {
+  const [prefix, setPrefix] = useState(networkPrefix);
+  const [isLoadingENSResolution, setIsLoadingENSResolution] = useState(false);
+
+  const throttle = useThrottle();
+
+  // checksum valid addresses
+  useEffect(() => {
+    if (isValidAddress(address) && !isChecksumAddress(address)) {
+      onChangeAddress(checksumAddress(address));
+    }
+  }, [address, onChangeAddress]);
+
+  // ENS resolution
+  useEffect(() => {
+    const resolveDomainName = async (ENSName: string) => {
+      try {
+        setIsLoadingENSResolution(true);
+        const address = (await getAddressFromDomain?.(ENSName)) as string;
+        onChangeAddress(address);
+        setPrefix(showNetworkPrefix ? networkPrefix : '');
+      } catch (e) {
+        onChangeAddress(ENSName);
+        setPrefix('');
+      } finally {
+        setIsLoadingENSResolution(false);
+      }
+    };
+
+    const isEnsName = isValidEnsName(address);
+
+    if (isEnsName && getAddressFromDomain) {
+      throttle(() => resolveDomainName(address));
+    }
+  }, [
+    address,
+    getAddressFromDomain,
+    networkPrefix,
+    showNetworkPrefix,
+    onChangeAddress,
+    throttle,
+  ]);
```

Comment:
Can this be merged into a single effect? I'm not sure if it's a good idea to run multiple effects that in the end call the same `onChangeAddress` function

Reply:
done! now we only have one useEffect!

each time we use onChangeAddress we checksumValidAddress first:

```
 onChangeAddress(checksumValidAddress(address));
```